src/core/qgscoordinatereferencesystem.cpp

```diff
@@ -2067,6 +2067,15 @@ bool QgsCoordinateReferenceSystem::readXml( const QDomNode &node )
 
       const QString wkt = srsNode.namedItem( QStringLiteral( "wkt" ) ).toElement().text();
       initialized = createFromWktInternal( wkt, description );
+
+      // the wkt can be a valid crs description but it can be an invalid crs
+      // this depends on the proj4 version
+      // a wkt provided by proj4 version >= 6 can be invalid for proj4 <= 5
+      // but the proj4 string provides a valid crs
+      if ( initialized && !isValid() )
```

Comment:
I don't understand this change -- as far as I can see, createFromWktInternal always returns d->mIsValid already from the expected return paths. Which return path in createFromWktInternal is being hit here?

Reply:
I have tested this code in QGIS Python Console:
```py
wkt = 'PROJCRS["unknown", BASEGEOGCRS["unknown", ... ] ]'
crs = QgsCoordinateReferenceSystem()
QgsCoordinateReferenceSystem.createFromWkt(crs, wkt)
# Returns True
crs.isValid()
# Returns False
```
If I do the same with a crs' wkt provided by proj4 version < 6, I have:
```py
wkt = 'PROJCS["unnamed", GEOGCS["WGS 84", ... ]'
crs = QgsCoordinateReferenceSystem()
QgsCoordinateReferenceSystem.createFromWkt(crs, wkt)
# Returns True
crs.isValid()
# Returns True
```

Reply:
Can you use a debugger and confirm which return path is being hit?

Reply:
The issue is line 872 https://github.com/qgis/QGIS/blob/master/src/core/qgscoordinatereferencesystem.cpp#L872 and in the other createFrom methods and probably in the way CRS definition cache is managed.

Reply:
Ah - in that case line 872 should read `return d->mIsValid`. Can you add a test for this too? (i.e. call createFromWkt twice with the same bad wkt string and test that both times it returns false).

Reply:
I have created a new PR with this fix and more https://github.com/qgis/QGIS/pull/39419